## 모델 구조와 가중치

In [2]:
from ultralytics import YOLO

model=YOLO('yolov8n.pt')
data_path='./datasets/data.yaml'
model.train(data=data_path, epochs=5, imgsz=640)

Ultralytics 8.3.191  Python-3.10.18 torch-2.6.0+cu124 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
engine\trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=./datasets/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=5, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=train, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, perspective=0.0, plots=True, pose=1

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x000001DBBF452D10>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          

In [4]:
from ultralytics import YOLO

model=YOLO('yolov8n.pt')
data_path='./datasets/data.yaml'
model.train(data=data_path, epochs=15, imgsz=640)

Ultralytics 8.3.191  Python-3.10.18 torch-2.6.0+cu124 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
engine\trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=./datasets/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=15, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=train2, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, perspective=0.0, plots=True, pose

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x000001DBBEA200D0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          

In [3]:
result=model.predict('sample.png')
result[0].show()


image 1/1 c:\githome\Object_detection\sample.png: 448x640 2 Papers, 1 Rock, 30.3ms
Speed: 2.6ms preprocess, 30.3ms inference, 3.2ms postprocess per image at shape (1, 3, 448, 640)


In [7]:
model=YOLO('runs/detect/train2/weights/best.pt')
data_path='./datasets/data.yaml'
metrics=model.val(data=data_path)
metrics

Ultralytics 8.3.191  Python-3.10.18 torch-2.6.0+cu124 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
Model summary (fused): 72 layers, 3,006,233 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access  (ping: 0.00.0 ms, read: 252.1139.0 MB/s, size: 27.3 KB)
val: Scanning C:\githome\Object_detection\datasets\valid\labels.cache... 576 images, 238 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 576/576  0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 36/36 11.6it/s 3.1s0.1s
                   all        576        400      0.943      0.931       0.95       0.76
                 Paper        132        139      0.953      0.928       0.96      0.779
                  Rock        121        141      0.936      0.941      0.942      0.739
              Scissors        116        120      0.941      0.923       0.95       0.76
Speed: 0.4ms preprocess, 2.6ms inference, 0.0ms loss, 0.6ms postprocess per image
Results saved to r

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x000001DBA2DCB280>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          

## 성능 비교 (yolov8n.pt vs best.pt)

- Precision: 모델이 "객체다"라고 예측했을 때, 실제 맞을 확률
- Recall: 실제 객체를 놓치지 않고 찾아낸 비율이 93.1%
- mAP@0.5: 모델이 얼마나 정확하게 객체를 찾아내고, 그 위치까지 잘 맞췄는지를 평균적으로 나타냄.


| 지표                 | yolov8n.pt | best.pt | 변화   |
|----------------------|-----------------|-------------------|--------|
| Precision (정밀도)   | 0.922           | 0.955             | 🔼 상승 |
| Recall (재현율)      | 0.855           | 0.918             | 🔼 상승 |
| mAP@0.5              | 0.916           | 0.950             | 🔼 상승 |
| mAP@0.5:0.95         | 0.691           | 0.759             | 🔼 상승 |
| Fitness (종합지표)   | 0.713           | 0.778             | 🔼 상승 |


## best_parameters.yaml

In [1]:
from ultralytics import YOLO
# YOLOv8 모델 로드
model = YOLO("yolov8n.pt")
# 하이퍼파라미터 탐색 수행
model.tune(
data="./datasets/data.yaml",
epochs=5,
imgsz=640,
iterations=5, # 탐색 횟수 (예: 25번의 실험)
val=True
)

Tuner: Initialized Tuner instance with 'tune_dir=runs\detect\tune2'
Tuner:  Learn about tuning at https://docs.ultralytics.com/guides/hyperparameter-tuning
Tuner: Starting iteration 1/5 with hyperparameters: {'lr0': 0.01, 'lrf': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'warmup_epochs': 3.0, 'warmup_momentum': 0.8, 'box': 7.5, 'cls': 0.5, 'dfl': 1.5, 'hsv_h': 0.015, 'hsv_s': 0.7, 'hsv_v': 0.4, 'degrees': 0.0, 'translate': 0.1, 'scale': 0.5, 'shear': 0.0, 'perspective': 0.0, 'flipud': 0.0, 'fliplr': 0.5, 'bgr': 0.0, 'mosaic': 1.0, 'mixup': 0.0, 'cutmix': 0.0, 'copy_paste': 0.0}
Saved runs\detect\tune2\tune_scatter_plots.png
Saved runs\detect\tune2\tune_fitness.png

Tuner: 1/5 iterations complete  (277.43s)
Tuner: Results saved to runs\detect\tune2
Tuner: Best fitness=0.71376 observed at iteration 1
Tuner: Best fitness metrics are {'metrics/precision(B)': 0.92147, 'metrics/recall(B)': 0.85509, 'metrics/mAP50(B)': 0.91551, 'metrics/mAP50-95(B)': 0.69134, 'val/box_loss': 0.92063, 'v

In [1]:
from ultralytics import YOLO
import yaml

model = YOLO("yolov8n.pt")
data_path = "./datasets/data.yaml"

# 튜닝 결과 하이퍼파라미터 파일 불러오기
with open("runs/detect/tune2/best_hyperparameters.yaml", "r", encoding="utf-8") as f:
    hyp_dict = yaml.safe_load(f)

# 최적 하이퍼파라미터로 학습
model.train(
    data=data_path,
    epochs=20,   # 원하는 만큼 학습 (예: 20, 50, 100)
    imgsz=640,
    batch=32,
    **hyp_dict
)


Ultralytics 8.3.191  Python-3.10.18 torch-2.6.0+cu124 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
engine\trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=32, bgr=0.0, box=5.24389, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.41947, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=./datasets/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.24473, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=20, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.01647, hsv_s=0.78431, hsv_v=0.30771, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01064, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=train12, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, perspec

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x0000027740B7FA00>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          

| 모델        | Precision (정밀도) | Recall (재현율) | mAP50 | mAP50-95 | Fitness |
|-------------|--------------------|----------------|-------|----------|---------|
| **yolov8n.pt**   | 0.918              | 0.904          | 0.935 | 0.721    | 0.745   |
| **best.pt**  | 0.927              | 0.912          | 0.947 | 0.752    | 0.768   |
| **최적 파라미터** | 0.939              | 0.926          | 0.954 | 0.769    | 0.787   |
